### ISSUE TITLE
edk2-stable202111: Black screen when doing single GPU passthrough (cause: TPM 2 related bugfix) (Bugzilla Bug 3771)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#10177]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 3771
Date: 2021-12-14T09:49:38+00:00
From: Stefan Berger &lt;<stefanb>&gt;
To:   unassigned &lt;<unassigned>&gt;
CC:   @ardbiesheuvel, @lgao4, @jyao1, @niruiyu

Last updated: 2022-01-04T21:57:13+00:00

**Comment by tianocore-issues** [comment]
## Comment 17985
Date: 2021-12-14 09:49:38 +0000
From: Stefan Berger &lt;<stefanb>&gt;

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): unassigned &lt;<unassigned>&gt;

Users of recent EDK2 package in ArchLinux are reporting a &apos;black screen when doing single GPU passthrough&apos;: https://bugs.archlinux.org/task/72991

The culprit has been found to be this patch here: https://github.com/tianocore/edk2/commit/b8675deaa819631db2667df63f89799fe65fc906

When this patch is reverted, then things work again.

The purpose of the TPM2 related changes was to fix bug #3499 ( https://bugzilla.tianocore.org/show_bug.cgi?id=3499 ). To keep the Physical Presence Opcode processing working, we needed to move their handling to an earlier location so that they can be handled before the TPM2 platform hierarchy password is disabled.

A possible cause for the failure is that EfiBootManagerDispatchDeferredImages () has something to do with it and in case of GPU does some initialization of the GPU that *now happens after* PlatformInitializeConsole(), thus leaving the screen blank.

**Comment by tianocore-issues** [comment]
## Comment 17986
Date: 2021-12-14 11:28:56 +0000
From: Stefan Berger &lt;<stefanb>&gt;

Due to the GPU case it looks like this is a block (in BdsPlatform.c) that needs to stay together in this order (which the TPM 2 related patch broke):


  //
  // Prevent further changes to LockBoxes or SMRAM.
  // Any TPM 2 Physical Presence Interface opcode must be handled before.
  //
  Handle = NULL;
  Status = gBS-&gt;InstallProtocolInterface (&amp;Handle,
                  &amp;gEfiDxeSmmReadyToLockProtocolGuid, EFI_NATIVE_INTERFACE,
                  NULL);
  ASSERT_EFI_ERROR (Status);

  //
  // Dispatch deferred images after EndOfDxe event and ReadyToLock
  // installation.
  //
  EfiBootManagerDispatchDeferredImages ();

  // We need to connect all trusted consoles for TCG PP. Here we treat all
  // consoles in OVMF to be trusted consoles.
  PlatformInitializeConsole (
    XenDetected() ? gXenPlatformConsole : gPlatformConsole);


We need to call this same block again when we need user interaction due to the physical presence opcode processing, which requires display and keyboard input. But the comment &apos;Any TPM 2 Physical Presence... must be handled before&apos; isn&apos;t going to let this work (in all cases) unless we somehow can prevent the disablement of the platform hierarchy (fixing the platform hierarchy bug was the purpose of the series) by possibly unregistering its end-of-dxe notification handler. (If that&apos;s possible at all). [ https://github.com/tianocore/edk2/blob/edk2-stable202111/SecurityPkg/Tcg/Tcg2PlatformDxe/Tcg2PlatformDxe.c#L74 ]. So we may have some requirements that get into each other&apos;s way: end-of-dxe signal to disable TPM2 hierarchy but GPU passthrough console can only be initialized after End-Of-DXE (likely due to security concerns) and only after that user interaction would be possible and after that only the execution of the physical presence opcode.


In the PPI code we currently have this here and have to have called the above block beforehand then:

  // Console for user interaction
  // We need to connect all trusted consoles for TCG PP. Here we treat all consoles in OVMF to be trusted consoles.
  EfiBootManagerConnectAllDefaultConsoles ();

https://github.com/tianocore/edk2/blob/edk2-stable202111/OvmfPkg/Library/Tcg2PhysicalPresenceLibQemu/DxeTcg2PhysicalPresenceLib.c#L597


Any suggestions would be appreciated...


Can we put the above block into a function and call if before the EfiBootManagerConnectAllDefaultConsoles ()? Somehow we need to prevent the end-of-dxe notification handler to do its job as well since otherwise some of the TPM 2 physical presence and menu commands won&apos;t work anymore.

**Comment by tianocore-issues** [comment]
## Comment 18092
Date: 2022-01-04 21:57:13 +0000
From: @lgao4

Include Ray, Jiewen and Ard for the discussion.
